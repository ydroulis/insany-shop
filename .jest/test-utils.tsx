import React, { ReactNode } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ProductsStoreProvider } from '../src/providers/productsStoreProvider';
import { CategoriesStoreProvider } from '../src/providers/categoriesStoreProvider';

const AllProviders = ({ children }: { children: ReactNode }) => {
  return (
    <ProductsStoreProvider>
      <CategoriesStoreProvider>{children}</CategoriesStoreProvider>
    </ProductsStoreProvider>
  );
};

const renderWithProviders = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllProviders, ...options });

export * from "@testing-library/react";
export { renderWithProviders as render };
