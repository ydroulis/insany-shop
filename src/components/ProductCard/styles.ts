'use client'

import styled from 'styled-components'
import Link from 'next/link'
import Image from 'next/image'

export const CardArticle = styled.article`
    display: flex;
    overflow: hidden;
    box-shadow: -6px 8px 20.4px 0px #0000000A;
    transition: transform 0.3s ease-in-out;
    color: inherit;

    height: 430px;

    background-color: #fff;
    border-radius: 19px;
    
    position: relative;

    p{
        margin: 0;
    }

    &:hover {
        transform: scale(1.02);
    }

    &:focus-visible {
        outline: 3px solid #2563eb;
        outline-offset: 4px;
        border-radius: 4px;
        transform: scale(1.02);
    }

    @media (min-width: 768px) {
        width: 340px;
        height: 520px;
    }

    @media (min-width: 1024px) {
        width: 300px;
        height: 480px;
    }

    @media (min-width: 1440px) {
        width: 356px;
        height: 520px;
        
    }
`

export const CardLink = styled(Link)`
    width: 100%;
    height: 198px;

    text-decoration: none;
`

export const CardImage = styled(Image)`
    width: 100%;
    object-fit: cover;

    @media (min-width: 768px) {
        height: 270px;
    }

    @media (min-width: 1024px) {
        height: 230px;
    }

    @media (min-width: 1440px) {
        height: 270px;
    }
`

export const Details = styled.div`
    display: flex;
    flex-direction: column;

    font-size: 0.875rem;

    padding: 0.75rem 1.313rem 1.75rem 1.313rem;

    position: relative;

    color: #000000;
`

export const Category = styled.span`
`

export const Title = styled.h2`
    font-weight: 700;
    font-size: 0.875rem;

    margin-top: 0.938rem;
`

export const Description = styled.p`
    margin-top: 0.313rem !important;
    text-align: left !important;
`

export const Price = styled.span`
    margin: 0.938rem 0rem;
    font-weight: 700;
    font-size: 1.5rem;
    color: #1B9847;
`

export const Stock = styled.span`
    position: absolute;
    right: 21px;
    bottom: 50px;
`

export const Rate = styled.div`
    display: flex;
    gap: 5px;
    align-items: baseline;
    position: absolute;

    top: 12px;
    right: 21px;

    font-weight: 700;
`

export const Action = styled.div`
    width: 305px;

    position: absolute;

    left: 21px;
    bottom: 28px;
    
    @media (min-width: 768px) {
        width: 300px;
        left: 21px;
        bottom: 28px;
    }

    @media (min-width: 1024px) {
        width: 260px;
        left: 21px;
        bottom: 18px;
    }

    @media (min-width: 1440px) {
        width: 314px;

        left: 21px;
        bottom: 28px;
    }
`