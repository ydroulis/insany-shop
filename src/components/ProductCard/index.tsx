import React from 'react';
import Button from '../Button';
import { GoStarFill } from "react-icons/go";

import * as S from './styles';
import { Product } from '@/types/Products';


const ProductCard: React.FC<Product> = ({
    id,
    name,
    description,
    price,
    image,
    category,
    stock,
    rating
}) => {
    return (
        <S.CardArticle>
            <S.CardLink
                href="#"
                aria-label={`Ver detalhes do produto ${name}`}
            >
                <S.CardImage
                    src={image}
                    width={200}
                    height={200}
                    placeholder='blur'
                    blurDataURL="..."
                    loading="lazy"
                    alt={`Imagem do produto ${name}`}
                />
                <S.Details>
                    <S.Rate>
                        <span>
                            <GoStarFill size={12} color='#FFE100' aria-hidden="true" focusable="false" />
                        </span>
                        <span>{rating}</span>
                    </S.Rate>
                    <S.Category>{category}</S.Category>
                    <S.Title>{name}</S.Title>
                    <S.Description>{description}</S.Description>
                    <S.Price>R$ {price.toFixed(2).replace('.', ',')}</S.Price>
                    <S.Stock className='stock' aria-label={`${stock} unidades em estoque`}>{stock} em estoque</S.Stock>
                    <Button
                        action={() => { }}
                        ariaLabel={`Adicionar ${name} ao carrinho`}
                        type="default"
                    >
                        Adicionar ao carrinho
                    </Button>
                </S.Details>
            </S.CardLink>
        </S.CardArticle>
    );
}

export default ProductCard;