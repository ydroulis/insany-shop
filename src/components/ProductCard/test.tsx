import React from "react";
import { render, screen, fireEvent } from "../../../.jest/test-utils";
import ProductCard from "./";
import type { Product } from "@/types/Products";
import type { CartActions } from "../../stores/cartStore";

const mockCartActions: Pick<CartActions, "addProductToCart"> = {
    addProductToCart: jest.fn(),
};

jest.mock("../../providers/cartStoreProvider", () => ({
    useCartStore: <T,>(selector: (state: typeof mockCartActions) => T): T =>
        selector(mockCartActions),
}));

jest.mock("../Button", () => {
    const MockButton = ({
        action,
        ariaLabel,
        children,
    }: {
        action: () => void;
        ariaLabel?: string;
        children: React.ReactNode;
    }) => (
        <button onClick={action} aria-label={ariaLabel}>
            {children}
        </button>
    );
    MockButton.displayName = "Button";
    return MockButton;
});

describe("<ProductCard />", () => {
    const mockProduct: Product = {
        id: 1,
        name: "Produto Teste",
        description: "Descrição do produto teste com mais detalhes",
        price: 99.99,
        image: "/fake-image.png",
        category: "Eletrônicos",
        stock: 10,
        rating: 4.5,
    };

    beforeEach(() => {
        jest.clearAllMocks();
        render(<ProductCard {...mockProduct} />);
    });

    it("should render the card link with correct aria-label", () => {
        const article = screen.getByRole("article", {
            name: /ver detalhes do produto produto teste/i,
        });
        expect(article).toBeInTheDocument();
    });

    it("should render product image with correct alt and src", () => {
        const image = screen.getByAltText(/imagem do produto produto teste/i);
        expect(image).toBeInTheDocument();
        expect(image).toHaveAttribute("src", expect.stringContaining("fake-image.png"));
    });

    it("should render product details correctly", () => {
        expect(screen.getByText(mockProduct.category as string)).toBeInTheDocument();
        expect(screen.getByRole("heading", { name: mockProduct.name })).toBeInTheDocument();
        expect(screen.getByText(mockProduct.description.slice(0, 58) + "...")).toBeInTheDocument();
        expect(
            screen.getByText(`R$ ${mockProduct.price.toFixed(2).replace(".", ",")}`)
        ).toBeInTheDocument();

        const stock = screen.getByText(/10 em estoque/i);
        expect(stock).toBeInTheDocument();
        expect(stock).toHaveAttribute("aria-label", "10 unidades em estoque");
    });

    it("should render the rating with number and svg star icon", () => {
        expect(screen.getByText(mockProduct.rating.toString())).toBeInTheDocument();
        const svgIcon = screen.getByTestId("star-icon");
        expect(svgIcon).toBeInTheDocument();
    });

    it("should render add to cart button and handle click with alert", () => {
        const alertSpy = jest.spyOn(window, "alert").mockImplementation(() => { });

        const button = screen.getByRole("button", {
            name: /adicionar produto teste ao carrinho/i,
        });

        fireEvent.click(button);

        expect(mockCartActions.addProductToCart).toHaveBeenCalledWith({
            id: mockProduct.id,
            name: mockProduct.name,
            description: mockProduct.description,
            price: mockProduct.price,
            image: mockProduct.image,
            stock: mockProduct.stock,
            items: 1,
        });

        expect(alertSpy).toHaveBeenCalledWith("Produto adicionado ao carrinho!");
        alertSpy.mockRestore();
    });

    it("should not break accessibility by nesting interactive elements", () => {
        const button = screen.getByRole("button", {
            name: /adicionar produto teste ao carrinho/i,
        });
        const article = screen.getByRole("article", {
            name: /ver detalhes do produto produto teste/i,
        });

        expect(article).toBeInTheDocument();
        expect(button).toBeInTheDocument();
    });

    it("should match snapshot", () => {
        const { container } = render(<ProductCard {...mockProduct} />);
        expect(container).toMatchSnapshot();
    });
});
