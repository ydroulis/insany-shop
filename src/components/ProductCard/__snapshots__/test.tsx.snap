// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<ProductCard /> should match snapshot 1`] = `
<div>
  <a
    class="sc-bRKDuR eiEcPF"
    href="#"
  >
    <article
      aria-label="Ver detalhes do produto Produto Teste"
      class="sc-hvigdm cOwQtH"
    >
      <img
        alt="Imagem do produto Produto Teste"
        class="sc-fhHczv jlpayb"
        data-nimg="1"
        decoding="async"
        height="200"
        loading="lazy"
        src="/_next/image?url=%2Ffake-image.png&w=640&q=75"
        srcset="/_next/image?url=%2Ffake-image.png&w=256&q=75 1x, /_next/image?url=%2Ffake-image.png&w=640&q=75 2x"
        style="color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
        width="200"
      />
      <div
        class="sc-ggWZvA dLqYET"
      >
        <div
          class="sc-kNOymR jdKBkv"
        >
          <span>
            <svg
              aria-hidden="true"
              color="#FFE100"
              fill="currentColor"
              focusable="false"
              height="12"
              stroke="currentColor"
              stroke-width="0"
              style="color: rgb(255, 225, 0);"
              viewBox="0 0 24 24"
              width="12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m12.672.668 3.059 6.197 6.838.993a.75.75 0 0 1 .416 1.28l-4.948 4.823 1.168 6.812a.75.75 0 0 1-1.088.79L12 18.347l-6.116 3.216a.75.75 0 0 1-1.088-.791l1.168-6.811-4.948-4.823a.749.749 0 0 1 .416-1.279l6.838-.994L11.327.668a.75.75 0 0 1 1.345 0Z"
              />
            </svg>
          </span>
          <span>
            4.5
          </span>
        </div>
        <span
          class="sc-dTvVRJ jvNhqL"
        >
          Eletrônicos
        </span>
        <h2
          class="sc-jwTyAe hxRbsY"
        >
          Produto Teste
        </h2>
        <p
          class="sc-hjsuWn dfzDzc"
        >
          Descrição do produto teste
        </p>
        <span
          class="sc-jJLAfE bKxnRq"
        >
          R$ 
          99,99
        </span>
        <span
          aria-label="10 unidades em estoque"
          class="sc-hwkwBN hQtOZw stock"
        >
          10
           em estoque
        </span>
        <button
          aria-label="Adicionar Produto Teste ao carrinho"
        >
          Adicionar ao carrinho
        </button>
      </div>
    </article>
  </a>
</div>
`;
